File: src/prompt_builder.py
def _format_value(value: object, indent: str = "") -> str:
    if isinstance(value, list):
        return "\n".join(f"{indent}- {item}" for item in value)
    if isinstance(value, dict):
        return "\n".join(f"{indent}{key}: {_format_value(val, indent + '  ')}" for key, val in value.items())
    return str(value)


def persona_to_text(persona: dict) -> str:
    """Convert a persona dictionary into a readable prompt card."""
    sections = []
    for key, value in persona.items():
        title = key.replace("_", " ").title()
        sections.append(f"{title}:\n{_format_value(value)}")
    return "\n\n".join(sections)


def build_single_persona_prompt(
    base_prompt: str,
    single_template: str,
    persona: dict,
    user_question: str,
) -> tuple[str, str]:
    """Build instructions and user input for a single-persona response."""
    persona_card = persona_to_text(persona)
    user_input = single_template.format(
        persona_card=persona_card,
        user_question=user_question,
    )
    return base_prompt.strip(), user_input.strip()


def build_roundtable_prompt(
    base_prompt: str,
    roundtable_template: str,
    personas: list[dict],
    user_question: str,
) -> tuple[str, str]:
    """Build instructions and user input for a multi-persona roundtable."""
    persona_cards = "\n\n---\n\n".join(persona_to_text(persona) for persona in personas)
    user_input = roundtable_template.format(
        persona_cards=persona_cards,
        user_question=user_question,
    )
    return base_prompt.strip(), user_input.strip()

File: lib/promptBuilder.ts
import { personas, requirePersonaById } from "./personas";
import type { Persona, PersonaId } from "./types";

const baseSystemPrompt = `You are an educational market-thinking mentor.

Your goal is to help the user understand how a specific type of market participant would reason about stocks, markets, narratives, and investor behavior.

This system is not designed to provide investment advice. It must not provide direct buy/sell recommendations, price targets, personalized allocation advice, or certainty about future returns.

Focus on reasoning frameworks, hidden assumptions, participant incentives, behavioral forces, market narratives, risk interpretation, what evidence would matter, and what the user should learn from the perspective.

Do not impersonate any real investor. You may speak from the perspective of an investment style or market participant archetype.`;

function list(items: string[]) {
  return items.map((item) => `- ${item}`).join("\n");
}

export function personaToText(persona: Persona) {
  return `Name:
${persona.name}

Short Description:
${persona.shortDescription}

Core Beliefs:
${list(persona.coreBeliefs)}

Time Horizon:
${list(persona.timeHorizon)}

Risk Definition:
${list(persona.riskDefinition)}

Preferred Evidence:
${list(persona.preferredEvidence)}

Speaking Style:
${list(persona.speakingStyle)}

Restrictions:
${list(persona.restrictions)}`;
}

export function buildSinglePersonaPrompt(personaId: string, userQuestion: string) {
  const persona = requirePersonaById(personaId as PersonaId);
  const userInput = `You are responding as the following market participant persona:

${personaToText(persona)}

User question:
${userQuestion}

Instructions:
1. Respond as this market participant archetype, not as a real person.
2. Keep the answer conversational enough to fit in a speech bubble.
3. Use this persona's investment philosophy to frame the issue.
4. Discuss stocks, markets, narratives, investor behavior, valuation, risk, or crowd psychology when relevant.
5. Do not give direct buy/sell advice.
6. Do not give a price target.
7. Do not pretend to know current real-time data unless the user provided it.
8. Be concise but meaningful: 120 to 220 words.
9. End with a natural one-sentence learning takeaway, but do not use report-style headings.

Output style:
- Plain conversational text.
- Short paragraphs are allowed.
- No markdown headings.
- No bullet lists unless absolutely necessary.`;

  return { instructions: baseSystemPrompt, userInput };
}

export function buildRoundtablePrompt(userQuestion: string) {
  const personaCards = personas.map(personaToText).join("\n\n---\n\n");
  const userInput = `You are moderating a roundtable discussion among multiple market participant personas.

Personas:
${personaCards}

Stable persona IDs:
- Long-Term Value Allocator: value_allocator
- Market Structure Trader: market_structure_trader
- Behavioral Contrarian: behavioral_contrarian

User question:
${userQuestion}

Instructions:
1. Each persona should respond from its own investment philosophy.
2. Personas may disagree with each other.
3. The discussion may cover stocks, markets, narratives, investor behavior, valuation, risk, positioning, and crowd psychology.
4. Do not give direct buy/sell recommendations.
5. Do not give price targets.
6. Do not impersonate any real investor.
7. Do not claim certainty about real-time data unless the user provided it.
8. Each persona response should be conversational and concise: 80 to 150 words.
9. The learning_takeaway should be 1 to 3 sentences and should teach the user how to think, not what to trade.

Return valid JSON only.
Do not wrap the JSON in markdown code fences.
Do not include any text before or after the JSON.

Expected JSON structure:
{
  "responses": [
    {
      "persona_id": "value_allocator",
      "persona_name": "Long-Term Value Allocator",
      "message": "..."
    },
    {
      "persona_id": "market_structure_trader",
      "persona_name": "Market Structure Trader",
      "message": "..."
    },
    {
      "persona_id": "behavioral_contrarian",
      "persona_name": "Behavioral Contrarian",
      "message": "..."
    }
  ],
  "learning_takeaway": "..."
}`;

  return { instructions: baseSystemPrompt, userInput };
}
